iroha_network/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+use std::{
+    convert::{TryFrom, TryInto},
+    error::Error,
+    io::prelude::*,
+    net::{IpAddr, SocketAddr, TcpListener, TcpStream},
+    str::FromStr,
+};
+
+pub mod prelude {
+    //! Re-exports important traits and types. Meant to be glob imported when using `iroha_network`.
+
+    #[doc(inline)]
+    pub use crate::{Request, Response, Stream};
+}
+
+pub const BUFFER_SIZE: usize = 2048;
+
+pub struct Network {
+    server_connection: TcpStream,
+}
+
+impl Network {
+    /// Creates a new client and connects it to the server.
+    pub fn connect(server_url: &str) -> Result<Network, Box<dyn Error>> {
+        Ok(Network {
+            server_connection: TcpStream::connect(server_url)?,
+        })
+    }
+
+    /// Sends `request` over established and persistent connection and gets `Response`.
+    pub fn send_request(&mut self, request: Request) -> Result<Response, Box<dyn Error>> {
+        Self::send_over_stream(&mut self.server_connection, request)
+    }
+
+    /// Establishes connection, sends `request` closes connection and returns `Response`.
+    pub fn send_request_to(server_url: &str, request: Request) -> Result<Response, Box<dyn Error>> {
+        let mut stream = TcpStream::connect(server_url)?;
+        Self::send_over_stream(&mut stream, request)
+    }
+
+    /// Listens on localhost on port `port`.
+    /// When there is an incoming connections passes it's `Stream` to `handler`.
+    pub fn listen<H>(port: u16, mut handler: H) -> Result<(), Box<dyn Error>>
```

Comment:
Let's use server_url instead of or in addition to port. Depending on deployment strategy we need to define different hosts.